odl/discr/discr_ops.py

```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+            the order. This requires left and right padding sizes to be
+            at most as large as the original array shape.
+
+            ``'order0'``: Extend constantly with the outmost values
+            (ensures continuity).
+
+            ``'order1'``: Extend with constant slope (ensures continuity of
+            the first derivative). This requires at least 2 values along
+            each axis where padding is applied.
+
+        pad_const : scalar, optional
+            Value to be used in the ``'constant'`` padding mode.
+
+        discr_kwargs: dict, optional
+            Keyword arguments passed to the `uniform_discr` constructor.
+        """
+        if not isinstance(domain, DiscreteLp):
+            raise TypeError('`domain` must be a `DiscreteLp` instance, '
+                            'got {!r}'.format(domain))
+
+        if not domain.is_uniform:
+            raise ValueError('`domain` is not uniformly discretized')
+
+        num_left = kwargs.pop('num_left', None)
+        discr_kwargs = kwargs.pop('discr_kwargs', {})
+
+        if range is None:
+            if ran_shp is None:
+                raise ValueError('either `range` or `ran_shp` must be '
+                                 'given')
+
+            num_left = normalized_scalar_param_list(
+                num_left, domain.ndim, param_conv=safe_int_conv)
+            self.__num_left = tuple(num_left)
+
+            range = _resize_discr(domain, ran_shp, num_left, discr_kwargs)
+
+        elif ran_shp is None:
+            if num_left is not None:
+                raise ValueError('`num_left` can only be combined with '
+                                 '`ran_shp`')
+
+            if not np.allclose(range.cell_sides, domain.cell_sides):
+                raise ValueError(
+                    'cell sides of domain and range differ significantly '
+                    '(difference {})'
+                    ''.format(range.cell_sides - domain.cell_sides))
+
+            self.__num_left = self._num_left_from_spaces(domain, range)
+
+        else:
+            raise ValueError('cannot combine `range` with `ran_shape`')
+
+        pad_mode = kwargs.pop('pad_mode', 'constant')
+        self.__pad_mode = str(pad_mode).lower()
+        if self.pad_mode not in _SUPPORTED_PAD_MODES:
+            raise ValueError("`pad_mode` '{}' not understood".format(pad_mode))
+
+        self.__pad_const = float(kwargs.pop('pad_const', 0.0))
+
+        # padding mode 'constant' with `pad_const != 0` is not linear
+        linear = (self.pad_mode != 'constant' or self.pad_const == 0.0)
+
+        super().__init__(domain, range, linear=linear)
+
+    @property
+    def num_left(self):
+        """Number of cells added to/removed from the left."""
+        return self.__num_left
+
+    @property
+    def pad_mode(self):
+        """Padding mode used by this operator."""
+        return self.__pad_mode
+
+    @property
+    def pad_const(self):
```

Comment:
I chose `pad_const` since it makes the connection to `'constant'` padding obvious.
We use `padding_value` in [other places](https://github.com/odlgroup/odl/blob/281d1d8763cc33e62a0e67268a5668fa033bdd95/odl/discr/diff_ops.py#L55-L56), which I think is inferior. `cval` is only shorter, not better. Not sure if one can speak of a "default" here. Even the names for the padding modes are not really standardized. For example, the mode `'constant'` in PyWavelets is the "replicate boundary values" mode, not "fill constant" as in SciPy. So I'd say we're free to choose whichever name we think is most appropriate.